tools/shared/src/main/scala/org/scalajs/core/tools/linker/CollectionsCompat.scala

```diff
@@ -0,0 +1,32 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.core.tools.linker
+
+import scala.collection.mutable
+
+private[linker] object CollectionsCompat {
+  implicit class MutableMapCompatOps[K, V](val __self: mutable.Map[K, V])
+      extends AnyVal {
+
+    // filterInPlace replaces retain
+    def filterInPlace(p: (K, V) => Boolean): Unit = {
+      // Believe it or not, this is the implementation of `retain` in 2.12.x:
```

Comment:
Hahahahaha, were you anticipating me complaining :P

Reply:
Actually I was anticipating myself looking back at this code and being appalled at the inefficiency of it.